File: leet_BS_278_first_bad.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        left, right, res = 1, n, n
        while left <= right:
            mid = (left + right) // 2
            if isBadVersion(mid):
                res = mid
                right = res - 1
            else:
                left = mid + 1
        return res

File: leet_BS_278_first_bad.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n)
    {
        int res = n;
        int L = 1;
        int R = n;
        
        while (L <= R)
        {
            int mid = (R - L) / 2 + L;
            if (isBadVersion(mid))
            {
                res = mid;
                R = res - 1;
            }
            else
                L = mid + 1;
        }
        return res;
    }
};
